### PR TITLE
EMMeTT support in SpeechLLM + tutorial for Lhotse Multimodal Dataloading

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

This PR extends NeMo and SpeechLLM with the following:
* EMMeTT (optimized training) support for SpeechLLM
* Support for joint audio and text training for `modular_models.py` and `modular_t5_models.py`
* Dedicated `estimate_token_bins.py` and `oomptimizer.py` in `scripts/speech_llm`.
* Dedicated data types for text and multimodal conversation data
* APIs for registering prompt formatting functions for tuples of (data_type, prompt_formatter_type)
* APIs for sequence length measurement (input, output, total) of various data types, seqlen filtering, and 1D+2D bucketing
* Generally a major cleanup of past messy half-solutions related to the above
* Tutorial about the changes related to data loading

**Collection**: multimodal

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/data/lhotse/cutset.py [hunk 1]
```diff
@@ -140,13 +175,14 @@ def read_dataset_config(config) -> tuple[CutSet, bool]:
...
-        "force_finite": config.force_finite,
-        "max_open_streams": config.max_open_streams,
+        "shuffle": config.get("shuffle", False),
+        "shard_seed": config.get("shard_seed", "trng"),
+        "text_field": config.get("text_field", "text"),
+        "lang_field": config.get("lang_field", "lang"),
+        "metadata_only": config.get("metadata_only", False),
+        "force_finite": config.get("force_finite", False),
+        "max_open_streams": config.get("max_open_streams", None),
+        "token_equivalent_duration": config.get("token_equivalent_duration", None),
```
Comment (pzelasko) [inline nemo/collections/common/data/lhotse/cutset.py:185]: TODO: document propagated attrs / include in the tutorial

File: nemo/collections/multimodal/speech_llm/models/modular_models.py [hunk 2]
```diff
@@ -1278,12 +1294,22 @@ def inference_step(self, dataloader_iter, mode):
...
+        text_batch = {k: v for k, v in batch.items() if k.startswith("text_")}
+        if audio_batch:
+            inputs_text = [self.tokenizer.ids_to_text(c.tolist()) for c in audio_batch['contexts']]
+            labels_text = [self.tokenizer.ids_to_text(a.tolist()) for a in audio_batch['answers']]
+            preds_text = [
+                self.tokenizer.ids_to_text(t[l.item() :][: data_cfg.get('tokens_to_generate')])
+                for t, l in zip(output['token_ids'], audio_batch['context_lengths'])
+            ]
+        else:
+            inputs_text = [self.tokenizer.ids_to_text(c.tolist()) for c in text_batch['text_context_ids']]
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_models.py:1307]: if doing pure text decoding, where do you update the inference_strategy to move away from the audio one used default in SALM?

File: nemo/collections/multimodal/speech_llm/models/modular_models.py [hunk 3]
```diff
@@ -1726,6 +1759,67 @@ def find_frozen_submodules(model):
             self.perception = self.trainer.strategy._setup_model(self.perception)
             self.perception = self.perception.cuda(torch.cuda.current_device())
 
+    def oomptimizer_schema(self, schema: str = "audio") -> dict:
+        """
+        Return a typing schema for optimal batch size calibration for various
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_models.py:1764]: can u elaborate a little bit on how to write up this function if we want to add other schema in future

File: nemo/collections/multimodal/speech_llm/models/modular_t5_models.py [hunk 4]
```diff
@@ -347,7 +348,7 @@ def prepare_llm_input(self, audio_batch):
         input_ids, input_length, labels, loss_mask = (
             audio_batch['contexts'],
             audio_batch['context_lengths'],
-            audio_batch['labels'],
+            audio_batch['answers'],
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:351]: any reason to update this ?

File: nemo/collections/multimodal/speech_llm/models/modular_t5_models.py [hunk 5]
```diff
@@ -367,47 +368,72 @@ def prepare_llm_input(self, audio_batch):
...
+        multimodal_output = {}
+
+        if 'audio_signal' in audio_batch:
+            encoder_input, attention_mask, enc_mask = self.prepare_llm_input(audio_batch)
+            # enc_input = speech and text prompt
+            # dec_input and label = text output label
+            b = audio_batch['answers'].shape[0]
+            device = audio_batch['answers'].device
+            dec_input = (
+                audio_batch['masked_answer_ids'] if 'masked_answer_ids' in audio_batch else audio_batch['answers']
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:392]: can remove masked_answer_ids for simplicity

File: nemo/collections/multimodal/speech_llm/models/modular_t5_models.py [hunk 6]
```diff
@@ -923,25 +985,42 @@ def inference_step(self, dataloader_iter, mode, dataloader_idx=0):
...
-            tokenizer=self.tokenizer,
-            bos_id=self.bos_id,
-        )
+        audio_batch = {k: v for k, v in batch.items() if not k.startswith("text_")}
+        text_batch = {k: v for k, v in batch.items() if k.startswith("text_")}
+        assert (
+            audio_batch or text_batch and not (audio_batch and text_batch)
+        ), f"Expecting only text or audio batch, got {len(text_batch)=} and {len(audio_batch)=}"
+
+        if 'audio_signal' in audio_batch:
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:997]: when will 'audio_signal' not in audio_batch?

File: nemo/collections/multimodal/speech_llm/models/modular_t5_models.py [hunk 7]
```diff
@@ -923,25 +985,42 @@ def inference_step(self, dataloader_iter, mode, dataloader_idx=0):
...
+            encoder_input, attention_mask, enc_mask = self.prepare_llm_input(audio_batch)
+            predicted_token_ids, log_probs = self.frozen_model.decode(
+                tokens_enc=None,
+                enc_mask=enc_mask,
+                num_tokens_to_generate=self._inference_config['tokens_to_generate'],
+                encoder_input=encoder_input,
+                tokenizer=self.tokenizer,
+                bos_id=self.bos_id,
+            )
+        if text_batch:
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:1009]: assert only one of audio_batch and text_batch can exisit

File: nemo/collections/multimodal/speech_llm/data/build_dataset.py [hunk 8]
```diff
@@ -131,16 +133,25 @@ def build_speechllm_dataloader(dataset, data_cfg, consumed_samples=0, is_predict
...
+                        if hasattr(ic, "manifest_filepath"):
+                            manifest_filepath.append(ic.manifest_filepath)
+                        else:
+                            assert ic.type == "txt_pair"
+                            manifest_filepath.append(ic.target_paths)
                 for cur_input_cfg in input_cfg[0].input_cfg:
                     conf = copy.deepcopy(data_cfg)
                     conf.input_cfg[0].input_cfg = [cur_input_cfg]
+                    OmegaConf.set_struct(conf, False)
+                    conf.force_finite = True
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/data/build_dataset.py:147]: will this cause a problem if we silently overwrite this flag here?

File: examples/multimodal/speech_llm/conf/bestow/modular_audio_gpt_config_cross_llama_lhotse_multi.yaml [hunk 9]
```diff
@@ -276,20 +278,20 @@ model:
       separate_prompt_and_response_with_newline: False
       truncation_field: "context" # Options: ['context', 'answer']
       index_mapping_dir: null # Path to a directory to write index mapping files.
-      prompt_template: "[INST]\n<<SYS>>\nPlease answer the following based on the previous speech feature.\n<</SYS>>\n\n{input}[/INST] {output}"
```
Comment (zhehuaichen) [inline examples/multimodal/speech_llm/conf/bestow/modular_audio_gpt_config_cross_llama_lhotse_multi.yaml:279]: document that this yaml has been updated to work with the prompt formatter based data processor

File: scripts/speech_llm/oomptimizer.py [hunk 10]
```diff
@@ -0,0 +1,578 @@
+#!/usr/bin/env python
```
Comment (zhehuaichen) [inline scripts/speech_llm/oomptimizer.py:1]: r.e. the problem of real memory consumption was larger than the result from oomptimizer ,  do we have any update ? if that was the case, how can we use oomptimizer?

**Review by zhehuaichen — COMMENTED** [review]
Thanks for the great work!

File: tests/collections/multimodal/test_emmett.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/multimodal/test_emmett.py:1]: I believe they are already in the CI under unit_tests CPU / GPU

File: nemo/collections/multimodal/speech_llm/models/modular_models.py (hunk 2 above)
Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/models/modular_models.py:1307]: All you need to do is to change the input_cfg. If the input type is nemo/nemo_tarred/etc then it will go the audio route. If it's txt_pair/etc it will go the text route. The dataset / model code is written to put those two in separate groups (text/audio) so it automatically does the right thing.

File: nemo/collections/multimodal/speech_llm/models/modular_models.py (hunk 3 above)
Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/models/modular_models.py:1764]: Sure. This is a description of what keys and values to expects in the mini-batch dict returned from dataloader. We expect different items for audio and text so you can select the one you want. We are using this inside `oomptimizer.py` to generate random data for checking OOM. I can write a separate tutorial for this.

File: nemo/collections/multimodal/speech_llm/data/build_dataset.py (hunk 8 above)
Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/data/build_dataset.py:147]: Since this is in predict/eval branch, we always need the dataloader to be finite, so no.

File: nemo/collections/multimodal/speech_llm/models/modular_t5_models.py (hunk 4 above)
Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:351]: Technically no, because it's unused anyway.

File: nemo/collections/multimodal/speech_llm/models/modular_t5_models.py (hunk 7 above)
Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:1009]: a few lines above there is 
```
        assert (
            audio_batch or text_batch and not (audio_batch and text_batch)
        ), f"Expecting only text or audio batch, got {len(text_batch)=} and {len(audio_batch)=}"
```

File: scripts/speech_llm/oomptimizer.py (hunk 10 above)
Comment (pzelasko) [reply scripts/speech_llm/oomptimizer.py:1]: Haven't looked into it further yet. At the moment I have to multiply the batch sizes obtained from oomptimizer by ~0.6 to  get it working.

**Review by zhehuaichen — APPROVED** [review]
lgtm. Can you document the SALM-GPT and SALM-T5 runs you have done somewhere in PR or a doc so that we can keep track or keep an eye on if not finished (and the baseline if you have)
